### PR TITLE
[YANG] Add route flow counter support

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
HLD: https://github.com/Azure/SONiC/pull/908

#### Why I did it

Add YANG models for route flow counters

#### How I did it

1. Add a new YANG definition for new counter group FLOW_CNT_ROUTE
2. Add a new YANG definition for new table FLOW_COUNTER_ROUTE_PATTERN_TABLE

#### How to verify it

New unit test cases

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106

### CODE REVIEW

File: src/sonic-yang-models/yang-models/sonic-flex_counter.yang [hunk 1]
```diff
@@ -211,8 +211,51 @@ module sonic-flex_counter {
...
+                key "vrf_name ip_prefix";
+
+                leaf vrf_name {
+                    type string {
+                        length 0..16;
+                    }
+                }
+
+                leaf ip_prefix {
+                    type union {
```
Comment (dgsudharsan) [inline src/sonic-yang-models/yang-models/sonic-flex_counter.yang:244]: Can we use the sonic-ip-prefix type defined here?
https://github.com/Azure/sonic-buildimage/blob/e88d7b3b934cea4a279fd99965c142076ea83c18/src/sonic-yang-models/yang-templates/sonic-types.yang.j2#L47

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: src/sonic-yang-models/yang-models/sonic-flex_counter.yang [hunk 2]
```diff
@@ -211,8 +211,51 @@ module sonic-flex_counter {
...
         /* end of container FLEX_COUNTER_TABLE */
+
+        container FLOW_COUNTER_ROUTE_PATTERN_TABLE {
+            description "Flow counter route pattern of config_db.json";
+
+            list PATTERN_LIST {
+
+                key "vrf_name ip_prefix";
+
+                leaf vrf_name {
```
[the remark this answers is not in the document]
Comment (Junchao-Mellanox) [reply src/sonic-yang-models/yang-models/sonic-flex_counter.yang:237]: There are two reasons that I did not add vrf_name reference here:
1. User can add a VRF that does not exist yet here. This feature can resolve the VRF name once the VRF is created.
2. The field vrf_name accept VRF name or VNET name.

[the remark this answers is not in the document]
Comment (dgsudharsan) [reply src/sonic-yang-models/yang-models/sonic-flex_counter.yang:237]: Can you please add these as comments in yang?

**Review by dgsudharsan — DISMISSED** [review]

**Comment by liat-grozovik** [comment]
@dgsudharsan could you please review following the changes request?

**Comment by zhangyanzhao** [comment]
@praveen-li  will help to review.

File: src/sonic-yang-models/yang-models/sonic-flex_counter.yang [hunk 3]
```diff
@@ -211,8 +211,53 @@ module sonic-flex_counter {
                 }
             }
 
+            container FLOW_CNT_ROUTE {
```
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-flex_counter.yang:214]: Kindly add configuration at:
https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/tests/files/sample_config_db.json

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/flex_counter.json [hunk 4]
```diff
@@ -97,8 +101,79 @@
...
+                "PATTERN_LIST": [
+                    {
+                        "vrf_name": "Vrf1",
+                        "ip_prefix": "1.1.1.0/24",
+                        "max_match_count": 30
+                    },
+                    {
+                        "vrf_name": "Vrf1",
+                        "ip_prefix": "2000::/64",
+                        "max_match_count": 30
```
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/flex_counter.json:133]: Please let me know, If config part of HLD is agreed upon already?

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/flex_counter.json [hunk 5]
```diff
@@ -97,8 +101,79 @@
...
+                }
+                ]
+            }
+        },
+        "sonic-flex_counter:sonic-flex_counter": {
+            "sonic-flex_counter:FLOW_COUNTER_ROUTE_PATTERN": {
+                "PATTERN_LIST": [
+                    {
+                        "vrf_name": "Vrf1",
+                        "ip_prefix": "1.1.1.0/24",
```
Comment (praveen-li) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/flex_counter.json:172]: Is this expected it,  i.e. the host bits are missing?

File: src/sonic-yang-models/yang-models/sonic-flex_counter.yang [hunk 6]
```diff
@@ -211,8 +211,53 @@ module sonic-flex_counter {
...
+                    1. User is allowed to configure a VRF that does not exist yet here, orchagent is designed to resolve the VRF name once the VRF is created.
+                    2. The field vrf_name accept both VRF name and VNET name.
+                    */
+                    type string {
+                        length 0..16;
+                    }
+                }
+
+                leaf ip_prefix {
+                    type stypes:sonic-ip-prefix;
```
Comment (praveen-li) [inline src/sonic-yang-models/yang-models/sonic-flex_counter.yang:249]: sonic-ip-prefix is used when host bits are present with '/prefixlen'.
i.e. 11.22.33.44/24 ,
inet one is used for 11.22.33.0/24
Hope the type aligned with your intention here.

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/flex_counter.json (hunk 5 above)
Comment (Junchao-Mellanox) [reply src/sonic-yang-models/tests/yang_model_tests/tests_config/flex_counter.json:170]: Yes, will use inet:ip-prefix.

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/flex_counter.json (hunk 4 above)
Comment (Junchao-Mellanox) [reply src/sonic-yang-models/tests/yang_model_tests/tests_config/flex_counter.json:133]: Yes, there is no comment for config part.

**Comment by Junchao-Mellanox** [comment]
@praveen-li, I have fixed all the comments, could you please review and sign-off?

**Review by dgsudharsan — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@prsunny could you please help to signoff?

File: src/sonic-yang-models/yang-models/sonic-flex_counter.yang (hunk 2 above)
[the remark this answers is not in the document]
Comment (prsunny) [reply src/sonic-yang-models/yang-models/sonic-flex_counter.yang:257]: can they remove the vrf/prefix if corresponding flow counter is configured?

**Comment by liat-grozovik** [comment]
@prsunny , @qiluo-msft  kindly reminder to signoff this PR

**Review by keboliu — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@prsunny , @qiluo-msft kindly reminder to signoff this PR

**Review by prsunny — APPROVED** [review]

**Pull request merged**